### ISSUE TITLE
'explicit' keyword is ignored to make implicit type conversions

### ISSUE THREAD

**Issue by yashnator** [wlav/cppyy#165]
Hi,

I used the 'explicit' keyword for the constructor to prevent unnecessary type conversions. However, cppyy seems to ignore this restriction and type-casts the object anyway. 

Here is a simple reproducer for the issue - 

````
import cppyy

cppyy.cppdef("""
class MyClass {
public:
   MyClass() = default;
   explicit MyClass(std::string const& str) {}
};

void foo(MyClass const& myObject) {
    std::cout << "Calling foo(MyClass const&)" << std::endl;
}
"""
)

print(cppyy.gbl.foo("This should work"))
````
It's mentioned in the documentation that a second iteration is done by cppyy to make the necessary conversions, and using the explicit keyword doesn't stop this iteration.

Any help would be appreciated,

Thanks

**Comment by wlav** [comment]
Yes, the `explicit` keyword is not exposed by Cling and thus ignored by cppyy. A pythonization that checks the type and rejects it, is the only workaround I can think of.

The point about a "second round" is that implicit conversion isn't a thing in Python, as everything is pass-by-reference. Hence implicit conversions are ignored initially, and only if no conversion matches, those that could have an implicit conversion are retried. (Implicit conversions are also very inefficient as the concept of a "temporary" isn't a Python thing either, so handling the temporary comes with lots of overhead.)

**Comment by hahnjo** [comment]
I just checked, Cling and ROOT meta expose this via the `kIsExplicit` bit in `TClingMethodInfo::Property` - this should be available in cppyy, no?

**Comment by wlav** [comment]
Indeed, missed that. Regardless, if this is for ROOT, you'll have to submit a bug report on the ROOT repo, not here, as the ROOT project contains a fork of cppyy, so any fixes here won't propagate.

**Comment by guitargeek** [comment]
Sure, we'll do that too!
